File: apps/kotlin/src/test/kotlin/chapter11/DatasetsTest.kt
package chapter11

import org.jetbrains.kotlinx.dataframe.api.dataFrameOf
import kotlin.test.Test
import kotlin.test.assertEquals

class PrepareSurvivedTest {
    @Test
    fun `客室クラスと年齢と男性かどうかを特徴量にし生存を正解ラベルにする`() {
        val df =
            dataFrameOf(
                "PassengerId" to listOf(1, 2),
                "Survived" to listOf(0, 1),
                "Pclass" to listOf(3, 1),
                "Sex" to listOf("male", "female"),
                "Age" to listOf(30.0, 40.0),
                "Fare" to listOf(8.0, 60.0),
            )

        val (x, t) = prepareSurvived(df)

        assertEquals(listOf("Pclass", "Age", "male"), x.columnNames())
        assertEquals(listOf(listOf<Any?>(3, 1), listOf<Any?>(30.0, 40.0), listOf<Any?>(1, 0)), x.columns().map { it.toList() })
        assertEquals(listOf("0", "1"), t)
    }

    @Test
    fun `年齢の欠損値を年齢の平均値で補完する`() {
        val df =
            dataFrameOf(
                "Survived" to listOf(0, 1, 1),
                "Pclass" to listOf(3, 1, 2),
                "Sex" to listOf("male", "female", "female"),
                "Age" to listOf(20.0, null, 40.0),
            )

        val (x, _) = prepareSurvived(df)

        assertEquals(listOf(20.0, 30.0, 40.0), x["Age"].toList())
    }
}

class PrepareCinemaTest {
    @Test
    fun `興行収入を正解ラベルにし特徴量の欠損値を平均値で補完する`() {
        val df =
            dataFrameOf(
                "cinema_id" to listOf(101, 102, 103),
                "SNS1" to listOf(100, null, 300),
                "SNS2" to listOf(500, 600, 700),
                "actor" to listOf(null, 20.0, 40.0),
                "original" to listOf(0, 1, 0),
                "sales" to listOf(9000, 9500, 10000),
            )

        val (x, t) = prepareCinema(df)

        assertEquals(listOf("SNS1", "SNS2", "actor", "original"), x.columnNames())
        assertEquals(
            listOf(listOf(100.0, 200.0, 300.0), listOf(500.0, 600.0, 700.0), listOf(30.0, 20.0, 40.0), listOf(0.0, 1.0, 0.0)),
            x.columns().map { it.toList() },
        )
        assertEquals(listOf(9000.0, 9500.0, 10000.0), t)
    }
}

File: apps/node/test/chapter11/datasets.test.ts
import { mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterAll, describe, expect, it } from "vitest";
import {
  loadCinema,
  loadSurvived,
  prepareCinema,
  prepareSurvived,
} from "../../src/chapter11/datasets.ts";

describe("CSV の読み込み", () => {
  const dir = mkdtempSync(join(tmpdir(), "chapter11-"));

  afterAll(() => {
    rmSync(dir, { recursive: true });
  });

  it("Survived の数値の列を数値にし空欄を null にする", () => {
    const csvFile = join(dir, "survived.csv");
    writeFileSync(
      csvFile,
      "\uFEFFPassengerId,Survived,Pclass,Sex,Age,Fare\n1,0,3,male,30,8\n2,1,1,female,,60\n",
    );

    expect(loadSurvived(csvFile)).toEqual([
      {
        PassengerId: "1",
        Survived: "0",
        Pclass: 3,
        Sex: "male",
        Age: 30,
        Fare: "8",
      },
      {
        PassengerId: "2",
        Survived: "1",
        Pclass: 1,
        Sex: "female",
        Age: null,
        Fare: "60",
      },
    ]);
  });

  it("cinema のすべての列を数値にし空欄を null にする", () => {
    const csvFile = join(dir, "cinema.csv");
    writeFileSync(
      csvFile,
      "cinema_id,SNS1,SNS2,actor,original,sales\n101,100,500,,0,9000\n",
    );

    expect(loadCinema(csvFile)).toEqual([
      {
        cinema_id: 101,
        SNS1: 100,
        SNS2: 500,
        actor: null,
        original: 0,
        sales: 9000,
      },
    ]);
  });
});

describe("prepareSurvived", () => {
  it("客室クラスと年齢と男性かどうかを特徴量にし生存を正解ラベルにする", () => {
    const rows = [
      { Survived: "0", Pclass: 3, Sex: "male", Age: 30 },
      { Survived: "1", Pclass: 1, Sex: "female", Age: 40 },
    ];

    const { x, t } = prepareSurvived(rows);

    expect(x).toEqual([
      { Pclass: 3, Age: 30, male: 1 },
      { Pclass: 1, Age: 40, male: 0 },
    ]);
    expect(t).toEqual(["0", "1"]);
  });

  it("年齢の欠損値を年齢の平均値で補完する", () => {
    const rows = [
      { Survived: "0", Pclass: 3, Sex: "male", Age: 20 },
      { Survived: "1", Pclass: 1, Sex: "female", Age: null },
      { Survived: "1", Pclass: 2, Sex: "female", Age: 40 },
    ];

    const { x } = prepareSurvived(rows);

    expect(x.map((row) => row.Age)).toEqual([20, 30, 40]);
  });
});

describe("prepareCinema", () => {
  it("興行収入を正解ラベルにし特徴量の欠損値を平均値で補完する", () => {
    const rows = [
      {
        cinema_id: 101,
        SNS1: 100,
        SNS2: 500,
        actor: null,
        original: 0,
        sales: 9000,
      },
      {
        cinema_id: 102,
        SNS1: null,
        SNS2: 600,
        actor: 20,
        original: 1,
        sales: 9500,
      },
      {
        cinema_id: 103,
        SNS1: 300,
        SNS2: 700,
        actor: 40,
        original: 0,
        sales: 10000,
      },
    ];

    const { x, t } = prepareCinema(rows);

    expect(x).toEqual([
      { SNS1: 100, SNS2: 500, actor: 30, original: 0 },
      { SNS1: 200, SNS2: 600, actor: 20, original: 1 },
      { SNS1: 300, SNS2: 700, actor: 40, original: 0 },
    ]);
    expect(t).toEqual([9000, 9500, 10000]);
  });
});
